File: Behavioral/state/state.py
class StateEngine:
    def on(self):
        pass

    def off(self):
        pass

    def speed_up(self):
        pass


class FighterJet:
    def __init__(self):
        self.speed = 0
        self.state = StateEngineOff(self)

    def on(self):
        self.state.on();

    def off(self):
        self.state.off();

    def speed_up(self):
        self.state.speed_up()


class StateEngineOff(StateEngine):
    def __init__(self, jet: FighterJet):
        self.jet = jet

    def on(self):
        print("ENGINE ON")
        self.jet.state = StateEngineOn(self.jet)

    def off(self):
        print("ENGINE ALREADY STOPPED")

    def speed_up(self):
        print("CANNOT SPEED UP STOPPED ENGINE")
        pass


class StateEngineOn(StateEngine):
    def __init__(self, jet: FighterJet):
        self.jet = jet

    def on(self):
        print("ENGINE ALREADY STARTED")

    def off(self):
        print("ENGINE OFF")
        self.jet.state = StateEngineOff(self.jet)

    def speed_up(self):
        print("SPEED UP BY 10")
        self.jet.speed += 10


fighter_f16 = FighterJet()
fighter_f16.off()
fighter_f16.speed_up()
fighter_f16.on()
fighter_f16.speed_up()
fighter_f16.on()
fighter_f16.off()
fighter_f16.off()

File: Behavioral/state/state.rs
trait StateEngine {
    fn on(&self) -> Option<Box<dyn StateEngine>>;
    fn off(&self) -> Option<Box<dyn StateEngine>>;
    fn speed_up(&self) -> u16;
}

struct StateEngineOff {

}

impl StateEngine for StateEngineOff {
    fn on(&self) -> Option<Box<dyn StateEngine>>{
        println!("ENGINE ON");
        Some(Box::new(StateEngineOn{}))
    }

    fn off(&self) -> Option<Box<dyn StateEngine>> {
        println!("ENGINE ALREADY STOPPED");
        None
    }

    fn speed_up(&self) -> u16 {
        println!("CANNOT SPEED UP STOPPED ENGINE");
        0
    }
}

struct StateEngineOn {

}

impl StateEngine for StateEngineOn {
    fn on(&self) -> Option<Box<dyn StateEngine>>{
        println!("ENGINE ALREADY STARTED");
        None
    }

    fn off(&self) -> Option<Box<dyn StateEngine>> {
        println!("ENGINE OFF");
        Some(Box::new(StateEngineOff{}))
    }

    fn speed_up(&self) -> u16 {
        println!("INCREASE SPEED BY 10");
        10
    }
}

pub struct FighterJet {
    state: Box<dyn StateEngine>,
    speed: u16
}

impl FighterJet {
    pub fn new() -> Self {
        FighterJet{state: Box::new(StateEngineOff{}), speed: 0}
    }

    fn on(&mut self) {
        if let Some(new_state) = self.state.on() {
            self.state = new_state;
        }
    }

    fn off(&mut self) {
        if let Some(new_state) = self.state.off() {
            self.state = new_state;
        }
    }

    fn speed_up(&mut self) {
        self.speed += self.state.speed_up();
    }
}


fn main() {
    let mut fighter_f16 = FighterJet::new();
    println!("{}", fighter_f16.speed);
    fighter_f16.off();
    fighter_f16.speed_up();
    fighter_f16.on();
    fighter_f16.speed_up();
    fighter_f16.on();
    fighter_f16.off();
    fighter_f16.off();
}
